server/src/main/java/org/elasticsearch/action/datastreams/autosharding/DataStreamAutoShardingService.java

```diff
@@ -0,0 +1,473 @@
+                    + "with previous autosharding event [{}] and oldest generation index [{}] being created on [{}] ",
+                dataStream.getName(),
+                dataStream.getAutoShardingEvent(),
+                oldestBackingIndex.getName(),
+                oldestIndexMeta.getCreationDate()
+            );
+
+            return new AutoShardingResult(
+                AutoShardingType.NO_CHANGE_REQUIRED,
+                writeIndex.getNumberOfShards(),
+                writeIndex.getNumberOfShards(),
+                TimeValue.ZERO,
+                writeIndexLoad
+            );
+        }
+
+        double maxIndexLoadWithinCoolingPeriod = getMaxIndexLoadWithinCoolingPeriod(
+            metadata,
+            dataStream,
+            writeIndexLoad,
+            reduceShardsCooldown,
+            nowSupplier
+        );
+
+        logger.debug(
+            "calculating the optimal number of shards for a potential decrease in number of shards for data stream [{}] with the"
+                + " max indexing load [{}] over the decrease shards cool down period",
+            dataStream.getName(),
+            maxIndexLoadWithinCoolingPeriod
+        );
+        long optimalReduceNumberOfShards = computeOptimalNumberOfShards(
+            minNumberWriteThreads,
+            maxNumberWriteThreads,
+            maxIndexLoadWithinCoolingPeriod
+        );
+
+        if (optimalReduceNumberOfShards < writeIndex.getNumberOfShards()) {
+            logger.debug(
+                "data stream autosharding service recommends decreasing the number of shards from [{}] to [{}] after [{}] cooldown for "
+                    + "data stream [{}]",
+                writeIndex.getNumberOfShards(),
+                optimalReduceNumberOfShards,
+                remainingReduceShardsCooldown,
+                dataStream.getName()
+            );
+
+            // we should reduce the number of shards
+            return new AutoShardingResult(
+                AutoShardingType.DECREASES_NUMBER_OF_SHARDS,
+                writeIndex.getNumberOfShards(),
+                Math.toIntExact(optimalReduceNumberOfShards),
+                remainingReduceShardsCooldown,
+                maxIndexLoadWithinCoolingPeriod
+            );
+        }
+
+        return new AutoShardingResult(
+            AutoShardingType.NO_CHANGE_REQUIRED,
+            writeIndex.getNumberOfShards(),
+            writeIndex.getNumberOfShards(),
+            TimeValue.ZERO,
+            writeIndexLoad
+        );
+    }
+
+    // Visible for testing
+    static long computeOptimalNumberOfShards(int minNumberWriteThreads, int maxNumberWriteThreads, double indexingLoad) {
+        return Math.max(
+            Math.min(roundUp(indexingLoad / ((double) minNumberWriteThreads / 2)), 3),
+            roundUp(indexingLoad / ((double) maxNumberWriteThreads / 2))
+        );
+    }
+
+    private static long roundUp(double value) {
+        return Math.round(Math.ceil(value));
```

Comment:
We can also do this as
```suggestion
        return (long) Math.ceil(value);
```
and avoid the call to `Math.round`, but really, whichever you prefer is fine, it's the most minor of nits :)